airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/DestinationTask.kt

```diff
@@ -0,0 +1,121 @@
+package io.airbyte.cdk.write
+
+import kotlin.math.min
+
+/**
+ * A task that can be executed by a DestinationRunner.
+ *
+ * concurrency: reports the desired concurrency:
+ * - id: a unique identifier for the task
+ * - perSync: the maximum number of tasks that can run concurrently in a sync
+ * - perStream: the maximum number of tasks that can run concurrently for a single stream
+ *
+ * If perSync or perStream is 0, the task will never be blocked by that concurrency limit.
+ * If concurrency is null, the task will always run.
+ *
+ * execute: executes the task and returns the next task to run.
+ */
+abstract class DestinationTask {
+    data class Concurrency(val id: String, val perSync: Int=0, val perStream: Int=0) {
+        fun available(): Int {
+            if (perStream > 0 && perSync > 0) {
+                return min(perStream, perSync)
+            } else if (perStream > 0) {
+                return perStream
+            } else if (perSync > 0) {
+                return perSync
+            } else {
+                return 0
+            }
+        }
+    }
+    open val concurrency: Concurrency? = null // null => always run
+
+    abstract fun execute(): DestinationTask
+}
+
+/**
+ * A task that controls the flow of execution. It is not itself executed.
+ * Rather, it guards, fans out, or resource-counts other tasks.
+ */
+abstract class ControlTask: DestinationTask() {
+    override fun execute(): DestinationTask {
+        throw IllegalStateException("Control tasks should not be executed")
+    }
+}
+
+/**
+ * A task that is affined to single stream.
+ */
+interface PerStream {
+    val stream: Stream
+}
+
+/**
+ * The implementing task will be provided with the reader end of
+ * a queue for its stream.
+ */
+interface RecordConsumer: PerStream {
+    var payload: Iterable<DestinationRecord>?
+    var endOfStream: Boolean
+    var forceFlush: Boolean
+}
+
+/**
+ * Control task that halts all task execution.
+ */
+class Done: ControlTask()
+
+/**
+ * Fan-out task that enqueues an instance of the underlying task for each stream.
+ */
+class ForEachStream(val taskFor: (Stream) -> DestinationTask): ControlTask()
+
+/**
+ * Fan-out task that enqueues an instance of the underlying task for the
+ * amount of available parallelism (ie, min(perSync, perStream)).
+ */
+class ForEachAvailable(val taskFor: (Int) -> DestinationTask): ControlTask()
```

Comment:
There's an ExactlyOnce guard. But you're suggesting more like "any time there's not at least one already running"?

Basically, each time we run T&D we've accomplished it for all batches up until then.

At the state machine level this can be a guard (name TBD).

At the standard interface level, this could just be "flag a batch as mergeable". So if mergeable is true, the framework dispatches the batch task behind the guard, and it will only get called each time a new batch is produced and it is not already running.

The implementor could just keep track of its high water mark in the destination and advance it each time it processed exactly one of those tasks at a time.

So (simplified)

```
Stage1: Batch
  + "stage1"
  + persisted=false
  + concurrency=4

Stage2: Batch
  + "stage2"
  + persisted=true
  + concurrency=4
  + exclusions("stage3")

Stage3: Batch
  + "stage3"
  + mergeable=true
  + concurrency=1
  + exclusions("stage2")
```

So: stage1 is processed 1-1 to make stage2, which triggers acking; stage2 is processed in parallel with itself but exclusive to stage 3. stage 3 has at most one task running, and all others are discarded until its complete, and when its done it advances its high watermark. During `closeStream` anything left unprocessed can be handled.

Reply:
nice, yeah that makes sense. I think the high watermark will probably get tricky - i.e. if an earlier batch of `copy into raw table` gets stuck for some reason, but we manage to execute T+D on a later batch, that's annoying... maybe this is another case where we really only need to track total record count

Reply:
We haven't talked at all yet about stuck tasks/timeouts/failure semantics. Kotlin has pretty good support for canceling running coroutines, but we'd still have to be able to handle failure modes.

So like: it try to commit a batch and fail => seems pretty easy, just return what you were given and let it get re-enqueued to try again

My one-only tasks gets stuck and maybe fails... I guess it's fine? As long as it gets canceled eventually? Since either A) the next one picks up from the same watermark or B) closeStream finalizes anyway?